File: quant-worker/backtest/optimizer.py
from __future__ import annotations

from dataclasses import dataclass
from importlib.metadata import version
from math import sqrt
from typing import Any

import numpy as np
import pandas as pd
from skfolio import RiskMeasure
from skfolio.optimization import (
    EqualWeighted,
    HierarchicalEqualRiskContribution,
    HierarchicalRiskParity,
    InverseVolatility,
    MaximumDiversification,
    MeanRisk,
    ObjectiveFunction,
    RiskBudgeting,
)


MIN_OBSERVATIONS = 30


@dataclass(frozen=True)
class OptimizerRequest:
    returns: pd.DataFrame
    method: str
    max_weight: float
    total_weight: float
    annualization_by_symbol: dict[str, int]
    target_return_pct: float | None = None
    target_volatility_pct: float | None = None
    risk_tolerance: float | None = None


def _estimator(request: OptimizerRequest, annualization: float):
    shared = {"max_weights": request.max_weight}
    if request.method == "equal_weight":
        return EqualWeighted()
    if request.method == "inverse_volatility":
        return InverseVolatility()
    if request.method == "minimum_variance":
        return MeanRisk(risk_measure=RiskMeasure.VARIANCE, **shared)
    if request.method == "maximum_sharpe":
        return MeanRisk(
            objective_function=ObjectiveFunction.MAXIMIZE_RATIO,
            risk_measure=RiskMeasure.STANDARD_DEVIATION,
            **shared,
        )
    if request.method == "target_return":
        if request.target_return_pct is None:
            raise ValueError("Target return is required.")
        return MeanRisk(
            risk_measure=RiskMeasure.VARIANCE,
            min_return=request.target_return_pct / 100 / annualization,
            **shared,
        )
    if request.method == "target_volatility":
        if request.target_volatility_pct is None or request.target_volatility_pct <= 0:
            raise ValueError("Target volatility is required.")
        return MeanRisk(
            objective_function=ObjectiveFunction.MAXIMIZE_RETURN,
            risk_measure=RiskMeasure.STANDARD_DEVIATION,
            max_standard_deviation=request.target_volatility_pct / 100 / sqrt(annualization),
            **shared,
        )
    if request.method == "risk_tolerance":
        if request.risk_tolerance is None or request.risk_tolerance <= 0:
            raise ValueError("Risk tolerance is required.")
        return MeanRisk(
            objective_function=ObjectiveFunction.MAXIMIZE_UTILITY,
            risk_measure=RiskMeasure.VARIANCE,
            risk_aversion=1 / request.risk_tolerance,
            **shared,
        )
    if request.method == "risk_parity":
        return RiskBudgeting(risk_measure=RiskMeasure.VARIANCE, **shared)
    if request.method == "most_diversified":
        return MaximumDiversification(**shared)
    if request.method == "minimum_correlation":
        return HierarchicalEqualRiskContribution(**shared)
    if request.method == "minimum_cvar":
        return MeanRisk(risk_measure=RiskMeasure.CVAR, **shared)
    if request.method == "hierarchical_risk_parity":
        return HierarchicalRiskParity(risk_measure=RiskMeasure.VARIANCE, **shared)
    raise ValueError(f"Unsupported optimizer method: {request.method}.")


def _normalized_weights(raw: Any, total: float, maximum: float) -> np.ndarray:
    weights = np.asarray(raw, dtype=float)
    if weights.ndim != 1 or not np.isfinite(weights).all() or (weights < -1e-8).any():
        raise ValueError("Optimizer returned invalid weights.")
    weights = np.maximum(weights, 0)
    if weights.sum() <= 0:
        raise ValueError("Optimizer returned zero weights.")
    weights = weights / weights.sum() * total
    while (weights > maximum + 1e-12).any():
        capped = weights >= maximum
        excess = float(np.maximum(weights[capped] - maximum, 0).sum())
        weights[capped] = maximum
        available = ~capped
        if not available.any() or float((maximum - weights[available]).sum()) + 1e-12 < excess:
            raise ValueError("Optimizer cannot satisfy the maximum-weight constraint.")
        base = weights[available]
        if float(base.sum()) <= 1e-12:
            addition = np.full(base.shape, excess / len(base))
        else:
            addition = excess * base / float(base.sum())
        weights[available] += addition
    return weights


def _basis_points(symbols: list[str], weights: np.ndarray, total_bps: int, max_bps: int):
    raw = weights / weights.sum() * total_bps
    result = np.floor(raw).astype(int)
    remainder = total_bps - int(result.sum())
    order = sorted(range(len(symbols)), key=lambda i: (-(raw[i] - result[i]), symbols[i]))
    while remainder:
        changed = False
        for index in order:
            if result[index] < max_bps:
                result[index] += 1
                remainder -= 1
                changed = True
                if remainder == 0:
                    break
        if not changed:
            raise ValueError("Unable to round allocation within the maximum-weight constraint.")
    return {symbol: int(result[index]) for index, symbol in enumerate(symbols)}


def _metrics(frame: pd.DataFrame, weights: np.ndarray, annualization: float) -> dict[str, float | None]:
    series = frame.to_numpy(dtype=float) @ (weights / weights.sum())
    mean = float(np.mean(series)) if len(series) else 0.0
    volatility = float(np.std(series, ddof=1)) if len(series) > 1 else 0.0
    equity = np.cumprod(1 + series)
    peaks = np.maximum.accumulate(equity) if len(equity) else np.array([1.0])
    max_drawdown = float(np.min(equity / peaks - 1)) if len(equity) else 0.0
    return {
        "expectedReturnPct": round(mean * annualization * 100, 2),
        "volatilityPct": round(volatility * sqrt(annualization) * 100, 2),
        "sharpe": None if volatility <= 1e-12 else round(mean / volatility * sqrt(annualization), 4),
        "maxDrawdownPct": round(max_drawdown * 100, 2),
    }


def optimize(request: OptimizerRequest) -> dict[str, Any]:
    frame = request.returns.copy()
    if not 1 <= frame.shape[1] <= 10:
        raise ValueError("Expected 1 to 10 assets.")
    if len(frame) < MIN_OBSERVATIONS:
        raise ValueError("Optimizer requires at least 30 aligned observations.")
    if frame.isna().any().any() or not np.isfinite(frame.to_numpy(dtype=float)).all():
        raise ValueError("Return series must be finite and aligned.")
    symbols = sorted(str(column) for column in frame.columns)
    frame = frame[symbols]
    if set(symbols) != set(request.annualization_by_symbol):
        raise ValueError("Annualization inputs do not match return symbols.")
    if not 0 < request.total_weight <= 1 or not 0 < request.max_weight <= 1:
        raise ValueError("Weight constraints must be between zero and one.")
    if request.max_weight * len(symbols) + 1e-9 < request.total_weight:
        raise ValueError("Maximum asset weight cannot satisfy the portfolio total.")

    split = max(MIN_OBSERVATIONS, int(len(frame) * 0.7))
    if split >= len(frame):
        split = len(frame) - 1
    train = frame.iloc[:split]
    test = frame.iloc[split:]
    annualization = float(np.mean([request.annualization_by_symbol[symbol] for symbol in symbols]))
    estimator = _estimator(request, annualization)
    estimator.fit(train)
    weights = _normalized_weights(estimator.weights_, request.total_weight, request.max_weight)
    total_bps = round(request.total_weight * 10_000)
    max_bps = round(request.max_weight * 10_000)
    weights_bps = _basis_points(symbols, weights, total_bps, max_bps)
    effective = np.array([weights_bps[symbol] / 10_000 for symbol in symbols], dtype=float)

    means = frame.mean()
    volatility = frame.std(ddof=1)
    correlation = frame.corr().fillna(0)
    asset_metrics = [
        {
            "symbol": symbol,
            "expectedReturnPct": round(
                float(means[symbol]) * request.annualization_by_symbol[symbol] * 100, 2
            ),
            "volatilityPct": round(
                float(volatility[symbol]) * sqrt(request.annualization_by_symbol[symbol]) * 100,
                2,
            ),
        }
        for symbol in symbols
    ]
    correlation_matrix = [
        {
            "symbol": symbol,
            "correlations": {
                other: round(1.0 if symbol == other else float(correlation.loc[symbol, other]), 4)
                for other in symbols
            },
        }
        for symbol in symbols
    ]
    full_metrics = _metrics(frame, effective, annualization)
    return {
        "method": request.method,
        "source": {
            "library": "skfolio",
            "version": version("skfolio"),
            "repository": "https://github.com/skfolio/skfolio",
            "directory": "awesome-quant: Portfolio Optimization & Risk Analysis",
            "license": "BSD-3-Clause",
        },
        "weightsBps": weights_bps,
        "expectedReturnPct": full_metrics["expectedReturnPct"],
        "volatilityPct": full_metrics["volatilityPct"],
        "sharpe": full_metrics["sharpe"],
        "observationCount": len(frame),
        "assetMetrics": asset_metrics,
        "correlationMatrix": correlation_matrix,
        "validation": {
            "split": "chronological_70_30",
            "trainObservationCount": len(train),
            "testObservationCount": len(test),
            "inSample": _metrics(train, effective, annualization),
            "outOfSample": _metrics(test, effective, annualization),
        },
        "warnings": [],
    }

File: src/lib/backtest/optimizer.ts
const MIN_OBSERVATIONS = 30;
const ITERATIONS = 500;
const GRADIENT_STEP = 1;

export type OptimizerInput = {
  returnsBySymbol: Record<string, number[]>;
  riskAversion: number;
  maxWeightBps: number;
  totalWeightBps?: number;
  periodsPerYear?: number;
};

export type OptimizerResult = {
  weightsBps: Record<string, number>;
  expectedReturnPct: number;
  volatilityPct: number;
  sharpe: number | null;
  observationCount: number;
  warnings: string[];
};

function mean(values: number[]) {
  return values.reduce((sum, value) => sum + value, 0) / values.length;
}

function covariance(left: number[], right: number[], leftMean: number, rightMean: number) {
  let total = 0;
  for (let index = 0; index < left.length; index += 1) {
    total += (left[index] - leftMean) * (right[index] - rightMean);
  }
  return total / Math.max(1, left.length - 1);
}

function dot(left: number[], right: number[]) {
  return left.reduce((sum, value, index) => sum + value * right[index], 0);
}

function projectCappedSimplex(values: number[], cap: number) {
  if (cap * values.length < 1 - 1e-12) {
    throw new Error("Maximum asset weight cannot satisfy the portfolio total.");
  }
  let lower = Math.min(...values.map((value) => value - cap));
  let upper = Math.max(...values);
  for (let iteration = 0; iteration < 80; iteration += 1) {
    const lambda = (lower + upper) / 2;
    const total = values.reduce(
      (sum, value) => sum + Math.min(cap, Math.max(0, value - lambda)),
      0,
    );
    if (total > 1) lower = lambda;
    else upper = lambda;
  }
  const lambda = (lower + upper) / 2;
  const projected = values.map((value) => Math.min(cap, Math.max(0, value - lambda)));
  const total = projected.reduce((sum, value) => sum + value, 0);
  return projected.map((value) => value / total);
}

function basisPoints(
  symbols: string[],
  weights: number[],
  totalWeightBps: number,
  maxWeightBps: number,
) {
  const raw = weights.map((weight) => weight * totalWeightBps);
  const rounded = raw.map(Math.floor);
  let remainder = totalWeightBps - rounded.reduce((sum, value) => sum + value, 0);
  const order = symbols
    .map((symbol, index) => ({ symbol, index, fraction: raw[index] - rounded[index] }))
    .sort(
      (left, right) => right.fraction - left.fraction || left.symbol.localeCompare(right.symbol),
    );
  while (remainder > 0) {
    const candidate = order.find((item) => rounded[item.index] < maxWeightBps);
    if (!candidate) throw new Error("Unable to repair optimized basis-point allocation.");
    rounded[candidate.index] += 1;
    remainder -= 1;
    candidate.fraction = -1;
    order.sort(
      (left, right) => right.fraction - left.fraction || left.symbol.localeCompare(right.symbol),
    );
  }
  return Object.fromEntries(symbols.map((symbol, index) => [symbol, rounded[index]]));
}

export function optimizeMeanVariance(input: OptimizerInput): OptimizerResult {
  const symbols = Object.keys(input.returnsBySymbol).sort();
  if (symbols.length < 1 || symbols.length > 10) throw new Error("Expected 1 to 10 assets.");
  if (!Number.isFinite(input.riskAversion) || input.riskAversion < 1 || input.riskAversion > 10) {
    throw new Error("Risk aversion must be between 1 and 10.");
  }
  const totalWeightBps = input.totalWeightBps ?? 10_000;
  if (!Number.isInteger(totalWeightBps) || totalWeightBps < 1 || totalWeightBps > 10_000) {
    throw new Error("Total optimized weight must be between 1 and 10,000 basis points.");
  }
  if (
    !Number.isInteger(input.maxWeightBps) ||
    input.maxWeightBps < 1 ||
    input.maxWeightBps > 10_000 ||
    input.maxWeightBps * symbols.length < totalWeightBps
  ) {
    throw new Error("Maximum asset weight cannot satisfy the portfolio total.");
  }
  const observations = input.returnsBySymbol[symbols[0]]?.length ?? 0;
  if (observations < MIN_OBSERVATIONS) {
    throw new Error("Optimizer requires at least 30 overlapping return observations.");
  }
  const series = symbols.map((symbol) => input.returnsBySymbol[symbol]);
  if (
    series.some(
      (values) => values.length !== observations || values.some((value) => !Number.isFinite(value)),
    )
  ) {
    throw new Error("Return series must be finite and aligned.");
  }

  const means = series.map(mean);
  const matrix = series.map((left, leftIndex) =>
    series.map((right, rightIndex) => covariance(left, right, means[leftIndex], means[rightIndex])),
  );
  const cap = input.maxWeightBps / totalWeightBps;
  let weights = projectCappedSimplex(Array(symbols.length).fill(1 / symbols.length), cap);
  for (let iteration = 0; iteration < ITERATIONS; iteration += 1) {
    const gradient = means.map(
      (expectedReturn, index) => expectedReturn - input.riskAversion * dot(matrix[index], weights),
    );
    weights = projectCappedSimplex(
      weights.map((weight, index) => weight + GRADIENT_STEP * gradient[index]),
      cap,
    );
  }

  const periodsPerYear = input.periodsPerYear ?? 252;
  if (!Number.isFinite(periodsPerYear) || periodsPerYear <= 0) {
    throw new Error("Periods per year must be positive.");
  }
  const periodReturn = dot(means, weights);
  const variance = Math.max(
    0,
    weights.reduce((total, weight, index) => total + weight * dot(matrix[index], weights), 0),
  );
  const annualVolatility = Math.sqrt(variance * periodsPerYear);
  const annualReturn = periodReturn * periodsPerYear;
  const singular = variance <= 1e-18;

  return {
    weightsBps: basisPoints(symbols, weights, totalWeightBps, input.maxWeightBps),
    expectedReturnPct: annualReturn * 100,
    volatilityPct: singular ? 0 : annualVolatility * 100,
    sharpe: singular ? null : annualReturn / annualVolatility,
    observationCount: observations,
    warnings: singular ? ["SINGULAR_COVARIANCE"] : [],
  };
}
